Add MALI data mode

matthewhoffman opened the pull request:
MALI data mode uses the MALI component but rather than making prognostic calculations it reads an ice thickness field from a previous simulation.  For now, it is assumed there is a single datamode file per mesh.  This could be expanded if needed.  A new variant of LISIO is also added: MPAS_DISLISIO_JRA1p5

matthewhoffman commented:
I placed a data mode file at `/global/cfs/cdirs/e3sm/inputdata/glc/mpasli/mpas.ais8to30km/ais_8to30km_datamode.nc` on Perlmutter for testing.  It currently contains monthly thickness fields from 2015-02-01 to 2058-05-01 but will updated with a more complete time range when I finish generating it.

Testing with:
```
./create_test --wait --walltime 0:30:00    ERS_Ld5.TL319_oQU240wLI_ais8to30.MPAS_DISLISIO_JRA1p5.pm-cpu_intel.mpaso-ocn_glcshelf
```

@jonbob , when I try to run this test, I get an error, because MALI is trying to read time `0001-01-01_00:00:00` on init.  Does that make sense to you?  What year does this JRA compset start with?  Or do you think this could be an issue with the calendar and streams handling in the glc driver?

xylar commented:
I believe JRA starts in 1958.

matthewhoffman commented:
Thanks, @jonbob. MALI does not currently have a way to cycle dates on forcing files, so we'll need to talk through if we need to add that capability or if we can set up a compset that uses consistent dates for what we need data mode to do.  The MALI simulation I am using starts in 2000 and ends in 2300 (but I will probably keep just the first 100 years, or maybe even less). 

@xylar and @cbegeman , what compset/forcing and year range do you think we should set up for this?  Do you think a JRA-forced G-case is the right configuration for the ocean?  If so, do you want to start in 1958 (JRA start date) or 2000 (MALI simulation actual start date)?  And does it matter if the start year is the actual start year or year 0?  Based on what we decide, I can adjust xtime in the MALI forcing file to match that convention, which I think is fine as we intend this to be schematic and not an actual ice-sheet history.

rljacob commented:
Why aren't you doing this by making a dglc model analogous to docn and datm ?

jonbob commented:
@rljacob -- it is one of the things we discussed, but wasn't sure who could do that

cbegeman commented:
It sounds like from what @jonbob is saying that there wouldn't be a problem starting in 2000 and not changing MALI's dates. JRA seems fine, but I think that we should make sure AIS ice runoff is removed in this test, which would not currently be the case (https://github.com/E3SM-Project/E3SM/blob/d4ca7d0606930f53f30eb7ff7dbb92ec83b83b84/components/mpas-ocean/bld/build-namelist#L728-L733).

xylar commented:
I concur, let's start at 2000 unless that causes unexpected issues.

matthewhoffman commented on components/mpas-ocean/cime_config/config_compsets.xml:
```diff
@@ -142,6 +142,11 @@
 
+  <compset>
+    <alias>GMPAS-JRA1p4-DIB-PISMF-DIS</alias>
+    <lname>2000_DATM%JRA-1p4-2018_SLND_MPASSI%DIB_MPASO%IBPISMFDATMFORCED_DROF%JRA-1p4-2018-AIS0ROF_MALI%DATA_SWAV</lname>
```
based on discussion, change to `JRA-1p4-2018-AIS0ICE` option?

matthewhoffman replied:
and switch from 1p4 to 1p5?

xylar replied:
Based on Slack discussion, wait for a PR that fixes things so runoff is removed from G-Cases as it is from B-Cases.  The, yes, switch to JRA1p5 but do not use any of the `-AIS0*` options so that MPASO takes care of the zeroing out.

matthewhoffman replied:
Update note: this PR will be rebased, and compset adjustments made, once https://github.com/E3SM-Project/E3SM/pull/6693 is merged.

xylar replied:
We need not to forget to do the rebase once https://github.com/E3SM-Project/E3SM/pull/6693 goes in, which I think will be soon.

matthewhoffman replied:
Note: I have rebased the PR and updated the compset per the discussion in this thread.

matthewhoffman replied:
My previous comment had failed to include the update from JRA 1p4 to 1p5.  I've made that change but when I try to test as before, I get an error:
```
./create_test  -q debug  --walltime 0:30:00    ERS_Ld5.TL319_oQU240wLI_ais8to30.GMPAS-JRA1p5-DIB-PISMF-DIS.pm-cpu_intel.mpaso-ocn_glcshelf

Using project from config_machines.xml: e3sm
create_test will do up to 1 tasks simultaneously
create_test will use up to 320 cores simultaneously
Creating test directory /pscratch/sd/h/hoffman2/e3sm_scratch/pm-cpu/ERS_Ld5.TL319_oQU240wLI_ais8to30.GMPAS-JRA1p5-DIB-PISMF-DIS.pm-cpu_intel.mpaso-ocn_glcshelf.20250121_201717_zwa1hh
RUNNING TESTS:
  ERS_Ld5.TL319_oQU240wLI_ais8to30.GMPAS-JRA1p5-DIB-PISMF-DIS.pm-cpu_intel.mpaso-ocn_glcshelf
Starting CREATE_NEWCASE for test ERS_Ld5.TL319_oQU240wLI_ais8to30.GMPAS-JRA1p5-DIB-PISMF-DIS.pm-cpu_intel.mpaso-ocn_glcshelf with 1 procs
Finished CREATE_NEWCASE for test ERS_Ld5.TL319_oQU240wLI_ais8to30.GMPAS-JRA1p5-DIB-PISMF-DIS.pm-cpu_intel.mpaso-ocn_glcshelf in 2.462909 seconds (PASS)
Starting XML for test ERS_Ld5.TL319_oQU240wLI_ais8to30.GMPAS-JRA1p5-DIB-PISMF-DIS.pm-cpu_intel.mpaso-ocn_glcshelf with 1 procs
Finished XML for test ERS_Ld5.TL319_oQU240wLI_ais8to30.GMPAS-JRA1p5-DIB-PISMF-DIS.pm-cpu_intel.mpaso-ocn_glcshelf in 0.456216 seconds (PASS)
Starting SETUP for test ERS_Ld5.TL319_oQU240wLI_ais8to30.GMPAS-JRA1p5-DIB-PISMF-DIS.pm-cpu_intel.mpaso-ocn_glcshelf with 1 procs
Finished SETUP for test ERS_Ld5.TL319_oQU240wLI_ais8to30.GMPAS-JRA1p5-DIB-PISMF-DIS.pm-cpu_intel.mpaso-ocn_glcshelf in 4.173617 seconds (FAIL). [COMPLETED 1 of 1]
    Case dir: /pscratch/sd/h/hoffman2/e3sm_scratch/pm-cpu/ERS_Ld5.TL319_oQU240wLI_ais8to30.GMPAS-JRA1p5-DIB-PISMF-DIS.pm-cpu_intel.mpaso-ocn_glcshelf.20250121_201717_zwa1hh
    Errors were:
        WARNING: Should be running with salinity restoring on!
                 But no file available for this grid.
        ERROR: /global/cfs/cdirs/fanssie/users/hoffman2/e3sm/E3SM-mali-data-mode/components/mpas-albany-landice/cime_config/buildnml /pscratch/sd/h/hoffman2/e3sm_scratch/pm-cpu/ERS_Ld5.TL319_oQU240wLI_ais8to30.GMPAS-JRA1p5-DIB-PISMF-DIS.pm-cpu_intel.mpaso-ocn_glcshelf.20250121_201717_zwa1hh FAILED, see above
```

@xylar  or @jonbob , can you help me diagnose what's causing this?

matthewhoffman commented:
> Why aren't you doing this by making a dglc model analogous to docn and datm ?

@rljacob , to follow up more on this question, we've discussed this option but feel doing this through the MALI component is critical for a few reasons.  This capability is serving as a quasi-coupling mode that helps us prepare the ocean model physics for an active, evolving MALI component.  By having MALI replay the ice thickness history from a previous MALI simulation, we ensure that everything about the configuration is identical to what MPAS-Ocean and the coupler will be seeing when we finally transition to fully coupled ice sheet simulations.  In particular, the ice-shelf basal melting functionality is in the coupler itself, as opposed to a component model, operating on the GLC grid at the OCN timestep, and is only enabled if GLC is active (and an ocean namelist is activated).  Ensuring that those complex settings are working correctly with a new DGLC component would be extra work with little benefit.  In particular, this does not seem worth the work because the data MALI mode is replaying previous _MALI simulation output_ as a stepping stone towards full coupling, as opposed to existing data components that are replaying _observations_ and meant to represent historical conditions accurately.  We don't anticipate extending this capability to historical reanalysis of ice-sheet thickness, in part because such reanalyses don't really exist and are not expected to anytime soon.

(Maybe this is more detail than you were asking for, but I wanted to have our offline discussions on this question documented.)

matthewhoffman commented:
Note for this PR: we discussed today that eventually we will need to create a MALI data mode version with the MALI 4km AIS mesh, but that will happen after the SORRM v3r4 ocean mesh based on the MALI initial condition is finalized.  At that point, we will run a standalone MALI 4km simulation to generate the data mode history.  That will all be part of a later, follow-on PR, and the 8km AIS data mode in the PR will still be useful for development and testing of MPAS-Ocean wetting and drying in the interim.

matthewhoffman commented:
After help from @jonbob , I got the start time and the JRA forcing time to be 2000-01-01.  I've also updated the MALI data file to include monthly thickness fields from 2000-01-01 to 2100-01-01.  

@xylar, this PR is now ready for review.  A run using the data mode can be created on Perlmutter with:
```
./create_test --wait  -q debug  --walltime 0:30:00    ERS_Ld5.TL319_oQU240wLI_ais8to30.MPAS_DISLISIO_JRA1p5.pm-cpu_intel.mpaso-ocn_glcshelf
```

If everyone is happy with it, I still need to put the data mode file on the LCRC server - right now it only exists locally on Perlmutter's inputdata space.

xylar commented on components/mpas-albany-landice/cime_config/buildnml:
```diff
@@ -285,6 +285,18 @@ def buildnml(case, caseroot, compname):
+            # in MALI data mode, read an input stream with ice thickness monthly
+            lines.append('    <stream name="data-mode-input"')
+            lines.append('            type="input"')
+            lines.append(f'            filename_template="{din_loc_root}/glc/mpasli/{glc_grid}/{grid_prefix}_datamode.nc"')
```
I think it would be a good idea to give this file a datestamp.  That way, it could at least be updated and it would be clear which version was expected.

matthewhoffman replied:
This is added in c09dbd60df19d97f3dda1667f709cdd34dd189df

xylar reviewed (approved):
This looks great to me!  I only have one suggestion, and it fits well with moving the data file over to LCRC.

I'd like to run a longer DISLISIO test but probably on Chrysalis once the data file is there, and that can happen once this is on E3SM.  I think it's ready to be moved over there.

matthewhoffman commented:
Since Xylar's review, I've made the following changes:
* rebased
* update compset definition
* corrected logic for how MALI_DYNAMIC is applied to nl options
* changed MALI_DYNAMIC attribute to MALI_PROGNOSTIC_MODE, which now takes on 3 values: PROGNOSTIC, STATIC, DATA

Same test as before passes:
```
./create_test --wait -q debug  --walltime 0:30:00    ERS_Ld5.TL319_oQU240wLI_ais8to30.MPAS_DISLISIO_JRA1p5.pm-cpu_intel.mpaso-ocn_glcshelf

PASS ERS_Ld5.TL319_oQU240wLI_ais8to30.MPAS_DISLISIO_JRA1p5.pm-cpu_intel.mpaso-ocn_glcshelf RUN
```

@xylar , let me know if you'd like any further changes or if I can move this over to the main repo now.  One thing that could receive attention is there are two compsets added in this PR in different files (allactive vs. mpas-ocean compset files) that are very similar compsets.  We probably don't need both.  @xylar , @jonbob , let me know if you have a preference of which to keep.  Similarly, there is not a compset that include MALI static mode - should we add one?

xylar commented on components/mpas-albany-landice/cime_config/buildnml:
```diff
@@ -70,6 +71,7 @@ def buildnml(case, caseroot, compname):
     elif glc_grid == 'mpas.ais8to30km':
         grid_date += '20221027'
         grid_prefix += 'ais_8to30km'
+        datamode_date += '20250121'
```
We will presumably add a similar file for `mpas.ais4to20km` in a follow-up PR?  I think that's the one that would be compatible with the SORRMr4 mesh I'm working on and the SORRMr5 mesh that would include a dry region.  But we could also make an Icos r8 (?) mesh that could be compatible with this lower res MALI mesh.  Forgive me if we've already discussed this and I've lost track.

matthewhoffman replied:
Yes, my plan was to add a datamode file for the recently updated 4km AIS mesh, most likely in a follow up PR.

xylar commented:
I would suggest we keep `GMPAS-JRA1p5-DIB-PISMF-DIS` and get rid of `MPAS_DISLISIO_JRA1p5`.  I think this is mostly a variant of a G-case with data MALI instead of no MALI.  That still seems like mostly a G-case, and those have traditionally lived in the mpas-ocean space rather than in the "allactive" space.  I also think the more verbose name is more helpful.  It isn't actually clear to me if `MPAS_DISLISIO_JRA1p5` would have `PISMF`, `DISMF` or snowcapping (though the last is hard to fathom), whereas that is clear in `GMPAS-JRA1p5-DIB-PISMF-DIS`.

xylar commented:
I think once you get rid of `MPAS_DISLISIO_JRA1p5` (assuming we're in agreement), this is ready for E3SM.

matthewhoffman commented:
Moved to https://github.com/E3SM-Project/E3SM/pull/6945

xylar commented:
I'll close this since it's moved.
